wsmaster/che-core-api-devfile/src/main/resources/schema/devfile.json

```diff
@@ -124,81 +133,331 @@
-          {
-            "properties": {
-              "type": {
-                "enum": [
-                  "kubernetes",
-                  "openshift"
-                ]
+          "required": ["type"]
+        },
+        "then": {
+          "allOf": [
+            {
+              "if": {
+                "properties": {
+                  "type": {
+                    "enum": [
+                      "cheEditor",
+                      "chePlugin"
+                    ]
+                  }
+                }
               },
-              "alias": {},
-              "reference": {},
-              "referenceContent": {},
-              "selector": {},
-              "entrypoints": {}
+              "then": {
+                "properties": {
+                  "type": {},
+                  "alias": {},
+                  "id": {
+                    "type": "string",
+                    "description": "Describes the component id. It has the following format: [{REGISTRY_URL}/]{plugin/editor ID}:{plugin/editor VERSION}, where `{REGISTRY_URL}/` is an optional part.",
+                    "pattern": "^((https?://)[a-zA-Z0-9_\\-\\./]+)?[a-zA-Z0-9_\\-\\.]{1,}:[a-zA-Z0-9_\\-\\.]{1,}$",
+                    "examples": [
+                      "org.eclipse.che.maven-jdk8:1.0.0",
+                      "https://che-plugin-registry.openshift.io/org.eclipse.che.maven-jdk8:1.0.0"
+                    ]
+                  },
+                  "memoryLimit": {
+                    "type": "string",
+                    "description": "Describes memory limit for the component. You can express memory as a plain integer or as a fixed-point integer using one of these suffixes: E, P, T, G, M, K. You can also use the power-of-two equivalents: Ei, Pi, Ti, Gi, Mi, Ki",
+                    "examples": [
+                      "128974848",
+                      "129e6",
+                      "129M",
+                      "123Mi"
+                    ]
+                  }
+                },
+                "required": [
+                  "id"
+                ],
+                "additionalProperties": false
+              }
             },
-            "anyOf": [
-              {
-                "required": ["reference"],
-                "additionalProperties": true
+            {
+              "if": {
+                "properties": {
+                  "type": {
+                    "enum": [
+                      "kubernetes",
+                      "openshift"
+                    ]
+                  }
+                }
               },
-              {
-                "required": ["referenceContent"],
-                "additionalProperties": true
-              }
-            ],
-            "additionalProperties": false
-          },
-          {
-            "properties": {
-              "type": {
-                "enum": [
-                  "dockerimage"
+              "then": {
+                "additionalProperties": false,
```

Comment:
In some entries additionalProperties goes after properties and in some before. Aligning order would make spec simpler to understand